docs/architecture/ADR-004-QGB_relayer_security.md

```diff
@@ -0,0 +1,82 @@
+# ADR 4: QGB Relayer security design
+
+## Changelog
+
+- {date}: {changelog}
+
+## Context
+
+The current QGB design requires relayers to relay everything in a perfect synchronous order, but the contracts do not.
+In fact, the QGB smart contract is designed to update the data commitments as follows:  
+- Receive a data commitment  
+- Check that the block height (nonce) is higher than the previously committed root  
+- Check if the data commitment is signed using the current valset _(this is the problematic check)_  
+- Then, other checks + commit
+So, if a relayer is up to date, it will submit data commitment and will pass the above checks.
+
+Now, if the relayer is missing some data commitments or valset updates, then it will start catching up the following way:  
+- Relay valset  
+- Keep relaying all data commitments that were signed using that valset  
+- If a new valset is found, check that: up to the block where the valset was changed, all the data commitments that happened during that period are relayed  
+- Relay the next valset  
+- And, so on.
+The problem with this approach is that there is constant risk of this happening for any relayer. Also, a malicious relayer, can target any honest QGB relayer in normal mode, or while catching up, and submit a valset before the honest relayer submits all the data commitments that were signed with the previous valset:
+-   Get the latest relayed valset
+-   Listen for new signed valsets
+-   Once a new valset is signed by 2/3 of the network, submit it immediately to be included in next block
+Then, this would create holes in the signatures as the honest relayer will no longer be able to submit the previous data commitments as they're not, in general, signed by the valset relayed by the malicious relayer. And the only solution is to jump to the ones that were signed with the current valset.
+
+## Alternative Approaches
+
+###  More  synchrony : Deploy the QGB contract with a data commitment window
+When deploying the QGB  contract,  also  set the data commitment window,  ie, the number of blocks between the `beginBlock` and `endBlock` of each data  commitment confirm.
+
+Then, update the QGB contract to check when receiving a new valset if the latest relayed data commitment height is >= new valset height - data commitment window.
+
+This also would mean adding, for example, a `DataCommitmentWindowConfirm` representing signatures of the validator set for a certain `DataCommitmentWindow`, since this latter can be updated using gov proposals.
```

Comment:
we probably want to update this portion of the ADR now to reflect our current plan to use universal nonce, but I like the wording here that describes the goal of adding synchrony. We have a data commitment window but the contract *is not* aware of it, which is an issue. How do we plan on using universal nonces, why does this fix the data commitment issue, and how does this fit in with the security assumptions the contract is already making?

Reply:
check now please? I added another section for the synchronous approach